### PR TITLE
[MRG+1] Added the possibility to use a Normalize class to modulate edge coloring 

### PR DESCRIPTION
**Pull request by demianw**
Added an optional Normalizer to the edge colouring of plotted graphs to be able to modulate edge colors.

### CODE REVIEW

**Comment by lesteve** [comment]
Just curious, what's your use case, making it easier to visually compare different connectome plots by using the same norm for all these plots?

Maybe having `edge_vmin` and `edge_vmax` arguments in plot_connectome would be easier to understand for the user.

**Comment by demianw** [comment]
My use case is connectomes which are based on connection probability or for instance plotting p-values over the edges. Also, integrating different diffusion-based connectivity measures which might require a log-scale colouring.

For all this, the use Normalize is the standard approach in matplotlib. I give you that vmin - vmax is a simplified interface but is does not suffice for my case and passing a normaliser class is a standard approach in matplotlib (for instance in imshow)

**Comment by GaelVaroquaux** [comment]
I am not very excited about exposing this super technical option to the
user while we are really striving to target simpler, less technical
users.

Is there a reason why you cannot process your data, for instance applying a
log, and then use a standard normalization?

**Comment by demianw** [comment]
@GaelVaroquaux, you mean just exposing vmin and vmax? 

The main reason would be that the whole matplotlib environment is geared towards using these for generating colorbars after. For instance, if I transform the data and then use it for plotting, it becomes super complicated to generate a colorbar with the right tick numbering as I have to go through every tick shown in the colorbar, apply the inverse transform and then change the ticks in the colorbar axis.

We've encountered different headaches in generating colorbars from other functions where this was not exposed.

**Comment by GaelVaroquaux** [comment]
OK, but trying to solve every corner case is going to blow up the
cyclomatic complexity of nilearn and make the API really hard to
understand (matplotlib is a good example of this). So we are taking the
"Steve Jobs" position on this and only catering to the 99% usecase.

Just to stress that this is not a theoretical argument but a practical
one, the recent few additions to the colorbar have made our codebase much
more complex and lead to a buggy release. We don't have the resources to
address these issues.

**Comment by demianw** [comment]
Sounds like an absolutely reasonable criteria. Would you rather I change it for the simpler vmin, vmax?
Nowadays the vmin-vmax behaviour is internally hardcoded and not documented.
It's in lines 266-270 of  nilearn/plotting/displays.py ( https://github.com/nilearn/nilearn/blob/028e89c6cc1be8ec9d35c2d358d1311f113ed1ad/nilearn/plotting/displays.py#L266 )
So it's hard to find out how the scale is being chosen (which is assuming an edge weight measure symmetric around 0).

**Comment by GaelVaroquaux** [comment]
> Sounds like an absolutely reasonable criteria. Would you rather I change it for
> the simpler vmin, vmax?

Yes, such an API is easily understandable by non technical users.

> Nowadays the vmin-vmax behaviour is internally hardcoded and not documented.
> It's in lines 266-270 of nilearn/plotting/displays.py

I agree that from a design perspective this is not ideal. I think that
exposing it and documentating it is a clear cut improvement.


Thanks

**Comment by demianw** [comment]
voila

**Comment by lesteve** [comment]
Sorry I completely missed the fact that you implemented the proposed changes. I'll take a closer look today.

File: nilearn/plotting/displays.py [hunk 1]
```diff
@@ -260,14 +261,25 @@ def _add_lines(self, line_coords, line_values, cmap, **kwargs):
...
+                will be used as minimum (multiplied by -1) and maximum
+                coloring levels.
             kwargs: dict
                 additional arguments to pass to matplotlib Line2D.
         """
         abs_line_values_max = np.abs(line_values).max()
-        norm = colors.Normalize(vmin=-abs_line_values_max,
-                                vmax=abs_line_values_max)
+        if vmin is None:
+            vmin = -abs_line_values_max
```
Comment (banilo) [inline nilearn/plotting/displays.py]: Just wondering, does it work if `vmin==vmax`, that is, all all connectivity strengths have the same value?

File: nilearn/plotting/displays.py [hunk 2]
```diff
@@ -1042,6 +1056,13 @@ def add_graph(self, adjacency_matrix, node_coords,
                 size(s) of the nodes in points^2.
             edge_cmap: colormap
                 colormap used for representing the strength of the edges.
+            edge_vmin: float, optional, default: None
+            edge_vmax: float, optional, default: None
+                If not None, either or both of these values will be used to
```
Comment (banilo) [inline nilearn/plotting/displays.py]: might not be explicit enough to convey the effect of `edge_vmin/max`on the plot...

File: nilearn/plotting/displays.py (hunk 1 above)
Comment (acil-bwh) [inline nilearn/plotting/displays.py]: This is replicates the previous implicit behaviour of the function where vmin and vmax where set. So the functional description still works. If vmin == vmax then all will be the same color which is the current matplotlib behaviour too.

Comment (demianw) [inline nilearn/plotting/displays.py]: This is replicates the previous implicit behaviour of the function where vmin and vmax where set. So the functional description still works. If vmin == vmax then all will be the same color which is the current matplotlib behaviour too

File: nilearn/plotting/displays.py (hunk 2 above)
Comment (demianw) [inline nilearn/plotting/displays.py]: I copied the current descritpion style in matplotlib. Do you have concrete suggestions?

File: nilearn/plotting/displays.py (hunk 1 above)
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: I think that the logic of this code will be surprising when vmin is none but not vmax.

I think that vmax should be set first,  and then vmin set to -vmax if none.

Comment (demianw) [inline nilearn/plotting/displays.py]: I agree. This type of clamping is very much an fMRI or co-variance matrix use-case so I rely on your expertise.

Comment (demianw) [inline nilearn/plotting/displays.py]: However, @GaelVaroquaux , what happens in your case if vmin is positive and vmax is None?

Comment (demianw) [inline nilearn/plotting/displays.py]: also the case where vmin is None and vmax is None will be quirky.

Comment (lesteve) [inline nilearn/plotting/displays.py]: > I think that vmax should be set first, and then vmin set to -vmax if none.

Agreed. That means:
* vmin and vmax are None: current behaviour, `vmax = np.max(np.abs(edge_data))`, `vmin = -vmax`
* vmin is None, vmax set: `vmin = -vmax`
* vmin set, vmax is None: `vmax = np.max(np.abs(edge_data))`, leave vmin as provided
* vmin and vmax set: leave vmax and vmin as provided

> However, @GaelVaroquaux , what happens in your case if vmin is positive and vmax is None?

See above. About these weird edge cases, I would be in favor of not trying to catch them all and just letting the matplotlib code do whatever it does in these cases.

Comment (demianw) [inline nilearn/plotting/displays.py]: I can do whatever you think is the best case however the lack of symmetry between cases 2 and 3 that you, @lesteve . I would either go with 
* vmin is None, vmax set: vmin = -vmax (throwing a ValueError if vmin is positive) 
* vmax is None, vmin set: vmax = -vmin (throwing a ValueError if vmax is negative)

or 

* vmin set, vmax is None: vmax = np.max(np.abs(edge_data))
* vmax set, vmin is None: vmin = -np.max(np.abs(edge_data))


But you guys have the last word

Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: >   • vmin is None, vmax set: vmin = -vmax (throwing a ValueError if vmin is
>     positive)
>   • vmax is None, vmin set: vmax = -vmin (throwing a ValueError if vmax is
>     negative)

> or

>   • vmin set, vmax is None: vmax = np.max(np.abs(edge_data))
>   • vmax set, vmin is None: vmin = -np.max(np.abs(edge_data))

First option is what I would expect.

Comment (lesteve) [inline nilearn/plotting/displays.py]: > First option is what I would expect.

Makes more sense than what I said indeed.

**Comment by GaelVaroquaux** [comment]
IMHO, the only remaining item to be done on this PR is the vmin / vmax behavior when the other is none (as discussed in inline comments). The rest is :+1: from me.

I'd love to get this in the 0.1.3 release.

**Comment by demianw** [comment]
done

**Comment by GaelVaroquaux** [comment]
LGTM, :+1: for merge.

Thanks @demianw !

**Comment by lesteve** [comment]
Merging, thanks a lot!

**Comment by GaelVaroquaux** [comment]
> Merged #556.

Hurry. Thanks again Demian

**Pull request merged**